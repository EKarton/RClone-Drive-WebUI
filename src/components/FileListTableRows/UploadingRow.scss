.uploading-row {
  &__status-text-wrapper {
    display: flex;
    column-gap: 1rem;
  }
}
