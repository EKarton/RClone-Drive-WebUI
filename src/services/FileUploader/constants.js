export const ActionTypes = Object.freeze({
  UPLOAD_FILE: 'UPLOAD_FILE',
  CANCEL_UPLOAD: 'CANCEL_UPLOAD',
});
