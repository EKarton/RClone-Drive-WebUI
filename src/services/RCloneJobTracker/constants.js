export const ActionTypes = Object.freeze({
  TRACK_NEW_JOB: 'TRACK_NEW_JOB',
  UNTRACK_NEW_JOB: 'UNTRACK_NEW_JOB',
});
