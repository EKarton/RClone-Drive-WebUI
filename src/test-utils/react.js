import { FileViewerProvider } from 'contexts/FileViewerStore';
import { RCloneInfoStateProvider } from 'contexts/RCloneInfoStore';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router';

const customRender = (
  component,
  { initialFileViewerState, initialRCloneInfoState } = {},
  { route = '/' } = {}
) => {
  const history = createMemoryHistory({ initialEntries: [route] });

  const renderedComponent = render(
    <RCloneInfoStateProvider defaultState={initialRCloneInfoState}>
      <FileViewerProvider defaultState={initialFileViewerState}>
        <Router history={history}>{component}</Router>
      </FileViewerProvider>
    </RCloneInfoStateProvider>
  );

  return {
    ...renderedComponent,
    history,
  };
};

export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { customRender, userEvent };
