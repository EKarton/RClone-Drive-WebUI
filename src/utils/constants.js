export const ICON_SIZE = Object.freeze({
  SMALL: 'small',
  MEDIUM: 'medium',
  LARGE: 'large',
});

export const StatusTypes = Object.freeze({
  LOADING: 'loading',
  SUCCESS: 'success',
  ERROR: 'error',
});

export const JobStatus = Object.freeze({
  ONGOING: 'ONGOING',
  SUCCESS: 'SUCCESS',
  ERROR: 'ERROR',
});

export const JobTypes = Object.freeze({
  UPLOAD_FILE: 'UPLOAD_FILE',
  MOVE_FILE: 'MOVE_FILE',
  MOVE_FOLDER: 'MOVE_FOLDER',
  RENAME_FILE: 'RENAME_FILE',
  RENAME_FOLDER: 'RENAME_FOLDER',
});

export const COLOR_MODE = Object.freeze({
  LIGHT: 'light',
  DARK: 'dark',
});

export const ImageMimeTypes = new Set(['image/jpeg', 'image/png']);
