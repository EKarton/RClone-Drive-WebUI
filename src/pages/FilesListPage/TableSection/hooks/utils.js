import { JobStatus, JobTypes } from 'utils/constants';

export function getFileNames(data) {
  return new Set(data.filter((file) => !file.IsDir).map((file) => file.Name));
}

export function getFolderNames(data) {
  return new Set(data.filter((file) => file.IsDir).map((file) => file.Name));
}

export function getUploadingJobs(jobs, remote, path) {
  return jobs.filter(
    (job) =>
      job.jobType === JobTypes.UPLOAD_FILE &&
      job.remote === remote &&
      job.dirPath.startsWith(path) &&
      job.status.value !== JobStatus.SUCCESS
  );
}

export function getUploadingFiles(jobs, remote, path) {
  return getUploadingJobs(jobs, remote, path).filter((job) => job.dirPath === path);
}

export function getUploadingFolders(jobs, remote, path) {
  return getUploadingJobs(jobs, remote, path).filter((job) => job.dirPath !== path);
}

export function getMovingJobs(jobs, remote, path) {
  const supportedJobTypes = new Set([
    JobTypes.MOVE_FILE,
    JobTypes.MOVE_FOLDER,
    JobTypes.RENAME_FILE,
    JobTypes.RENAME_FOLDER,
  ]);

  return jobs.filter((job) => {
    const correctJob = supportedJobTypes.has(job.jobType);
    const inProgress = job.status.value === JobStatus.ONGOING;

    if (!correctJob || !inProgress) {
      return false;
    }

    const isMatchSrc = job.src?.remote === remote && job.src?.dirPath === path;
    const isMatchTarget = job.target?.remote === remote && job.target?.dirPath === path;

    return correctJob && inProgress && (isMatchSrc || isMatchTarget);
  });
}
