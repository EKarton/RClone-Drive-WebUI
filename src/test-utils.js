import { FileViewerProvider, initialState } from 'contexts/FileViewerStore';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router';

const customRender = (
  component,
  { initialFileViewerState = initialState } = {},
  { route = '/' } = {}
) => {
  const history = createMemoryHistory({ initialEntries: [route] });

  return render(
    <FileViewerProvider defaultState={initialFileViewerState}>
      <Router history={history}>{component}</Router>
    </FileViewerProvider>
  );
};

export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { customRender, userEvent };
