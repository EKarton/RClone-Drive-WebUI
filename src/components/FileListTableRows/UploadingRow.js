import DescriptionIcon from '@mui/icons-material/Description';
import FolderIcon from '@mui/icons-material/Folder';
import CircularProgress from '@mui/material/CircularProgress';
import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';
import { BehaviorSubject } from 'rxjs';
import { ICON_SIZE, JobStatus } from 'utils/constants';
import BaseRow from './BaseRow';
import './UploadingRow.scss';

export default function UploadingRow({ file, iconSize }) {
  const [status, setStatus] = useState(null);

  useEffect(() => {
    const subscriber = file.status.subscribe((val) => {
      setStatus(val);
    });

    return () => {
      subscriber.unsubscribe();
    };
  }, [file.status]);

  const getIcon = () => {
    if (file.isDirectory) {
      return <FolderIcon fontSize={iconSize} />;
    }

    return <DescriptionIcon fontSize={iconSize} />;
  };

  const getUploadStatusText = () => {
    if (status === JobStatus.SUCCESS) {
      return 'Uploaded just now';
    }

    if (status === JobStatus.ERROR) {
      return 'Upload failed';
    }

    return (
      <div className="uploading-row__status-text-wrapper">
        <CircularProgress size={20} />
        <span>Uploading...</span>
      </div>
    );
  };

  return (
    <BaseRow
      fileIcon={getIcon()}
      fileName={file.name}
      dateModified={getUploadStatusText()}
      fileSize="-"
    />
  );
}

UploadingRow.propTypes = {
  file: PropTypes.shape({
    name: PropTypes.string,
    isDirectory: PropTypes.bool,
    status: PropTypes.instanceOf(BehaviorSubject),
  }),
  iconSize: PropTypes.oneOf(Object.values(ICON_SIZE)),
};
