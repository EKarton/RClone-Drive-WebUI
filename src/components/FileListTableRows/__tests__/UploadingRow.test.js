import { BehaviorSubject } from 'rxjs';
import { JobStatus } from 'utils/constants';
import { render, act } from 'test-utils/react';
import UploadingRow from '../UploadingRow';

describe('UploadingRow', () => {
  it('should match snapshot given a file', () => {
    const { baseElement } = render(<UploadingRow file={createFile()} />);

    expect(baseElement).toMatchSnapshot();
  });

  it('should match snapshot given a directory', () => {
    const fileObj = {
      ...createFile(),
      name: '2021',
      isDirectory: true,
    };

    const { baseElement } = render(<UploadingRow file={fileObj} />);

    expect(baseElement).toMatchSnapshot();
  });

  it.each(Object.values(JobStatus))(
    'should match snapshot given job status is %s',
    (status) => {
      const fileObj = {
        ...createFile(),
        status: new BehaviorSubject(status),
      };

      const { baseElement } = render(<UploadingRow file={fileObj} />);

      expect(baseElement).toMatchSnapshot();
    }
  );

  it('should update snapshot when file status updates', () => {
    const fileObj = createFile();

    const { baseElement } = render(<UploadingRow file={fileObj} />);

    act(() => fileObj.status.next(JobStatus.SUCCESS));

    expect(baseElement).toMatchSnapshot();
  });

  const createFile = () => ({
    name: 'dog.png',
    isDirectory: false,
    status: new BehaviorSubject(JobStatus.ONGOING),
  });
});
