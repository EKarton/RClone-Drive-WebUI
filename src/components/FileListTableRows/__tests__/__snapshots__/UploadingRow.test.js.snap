// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadingRow should match snapshot given a directory 1`] = `
<body>
  <div>
    <tr
      class="MuiTableRow-root filelist-table-base-row css-1q1u3t4-MuiTableRow-root"
      tabindex="0"
    >
      <td
        class="MuiTableCell-root MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-bsph2u-MuiTableCell-root"
      >
        <div
          class="filelist-table-base-row__file-cell"
          data-testid="2021"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="FolderIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2h-8l-2-2z"
            />
          </svg>
          <div>
            2021
          </div>
        </div>
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-bsph2u-MuiTableCell-root"
      >
        <div
          class="uploading-row__status-text-wrapper"
        >
          <span
            class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
            role="progressbar"
            style="width: 20px; height: 20px;"
          >
            <svg
              class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </span>
          <span>
            Uploading...
          </span>
        </div>
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-bsph2u-MuiTableCell-root"
      >
        -
      </td>
    </tr>
  </div>
</body>
`;

exports[`UploadingRow should match snapshot given a file 1`] = `
<body>
  <div>
    <tr
      class="MuiTableRow-root filelist-table-base-row css-1q1u3t4-MuiTableRow-root"
      tabindex="0"
    >
      <td
        class="MuiTableCell-root MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-bsph2u-MuiTableCell-root"
      >
        <div
          class="filelist-table-base-row__file-cell"
          data-testid="dog.png"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DescriptionIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8l-6-6zm2 16H8v-2h8v2zm0-4H8v-2h8v2zm-3-5V3.5L18.5 9H13z"
            />
          </svg>
          <div>
            dog.png
          </div>
        </div>
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-bsph2u-MuiTableCell-root"
      >
        <div
          class="uploading-row__status-text-wrapper"
        >
          <span
            class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
            role="progressbar"
            style="width: 20px; height: 20px;"
          >
            <svg
              class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </span>
          <span>
            Uploading...
          </span>
        </div>
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-bsph2u-MuiTableCell-root"
      >
        -
      </td>
    </tr>
  </div>
</body>
`;

exports[`UploadingRow should match snapshot given job status is ERROR 1`] = `
<body>
  <div>
    <tr
      class="MuiTableRow-root filelist-table-base-row css-1q1u3t4-MuiTableRow-root"
      tabindex="0"
    >
      <td
        class="MuiTableCell-root MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-bsph2u-MuiTableCell-root"
      >
        <div
          class="filelist-table-base-row__file-cell"
          data-testid="dog.png"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DescriptionIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8l-6-6zm2 16H8v-2h8v2zm0-4H8v-2h8v2zm-3-5V3.5L18.5 9H13z"
            />
          </svg>
          <div>
            dog.png
          </div>
        </div>
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-bsph2u-MuiTableCell-root"
      >
        Upload failed
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-bsph2u-MuiTableCell-root"
      >
        -
      </td>
    </tr>
  </div>
</body>
`;

exports[`UploadingRow should match snapshot given job status is ONGOING 1`] = `
<body>
  <div>
    <tr
      class="MuiTableRow-root filelist-table-base-row css-1q1u3t4-MuiTableRow-root"
      tabindex="0"
    >
      <td
        class="MuiTableCell-root MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-bsph2u-MuiTableCell-root"
      >
        <div
          class="filelist-table-base-row__file-cell"
          data-testid="dog.png"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DescriptionIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8l-6-6zm2 16H8v-2h8v2zm0-4H8v-2h8v2zm-3-5V3.5L18.5 9H13z"
            />
          </svg>
          <div>
            dog.png
          </div>
        </div>
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-bsph2u-MuiTableCell-root"
      >
        <div
          class="uploading-row__status-text-wrapper"
        >
          <span
            class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
            role="progressbar"
            style="width: 20px; height: 20px;"
          >
            <svg
              class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </span>
          <span>
            Uploading...
          </span>
        </div>
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-bsph2u-MuiTableCell-root"
      >
        -
      </td>
    </tr>
  </div>
</body>
`;

exports[`UploadingRow should match snapshot given job status is SUCCESS 1`] = `
<body>
  <div>
    <tr
      class="MuiTableRow-root filelist-table-base-row css-1q1u3t4-MuiTableRow-root"
      tabindex="0"
    >
      <td
        class="MuiTableCell-root MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-bsph2u-MuiTableCell-root"
      >
        <div
          class="filelist-table-base-row__file-cell"
          data-testid="dog.png"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DescriptionIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8l-6-6zm2 16H8v-2h8v2zm0-4H8v-2h8v2zm-3-5V3.5L18.5 9H13z"
            />
          </svg>
          <div>
            dog.png
          </div>
        </div>
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-bsph2u-MuiTableCell-root"
      >
        Uploaded just now
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-bsph2u-MuiTableCell-root"
      >
        -
      </td>
    </tr>
  </div>
</body>
`;

exports[`UploadingRow should update snapshot when file status updates 1`] = `
<body>
  <div>
    <tr
      class="MuiTableRow-root filelist-table-base-row css-1q1u3t4-MuiTableRow-root"
      tabindex="0"
    >
      <td
        class="MuiTableCell-root MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-bsph2u-MuiTableCell-root"
      >
        <div
          class="filelist-table-base-row__file-cell"
          data-testid="dog.png"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DescriptionIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8l-6-6zm2 16H8v-2h8v2zm0-4H8v-2h8v2zm-3-5V3.5L18.5 9H13z"
            />
          </svg>
          <div>
            dog.png
          </div>
        </div>
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-bsph2u-MuiTableCell-root"
      >
        Uploaded just now
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-bsph2u-MuiTableCell-root"
      >
        -
      </td>
    </tr>
  </div>
</body>
`;
